cloud/aws/aws_env.cc

```diff
@@ -213,54 +214,59 @@ void AwsEnv::Shutdown() { Aws::ShutdownAPI(Aws::SDKOptions()); }
 
 
 // The factory method for creating an S3 Env
-Status AwsEnv::NewAwsEnv(Env* base_env, const CloudEnvOptions& cloud_options,
+Status AwsEnv::NewAwsEnv(const std::shared_ptr<FileSystem>& base_fs,
+                         const CloudEnvOptions& cloud_options,
                          const std::shared_ptr<Logger>& info_log,
                          CloudEnv** cenv) {
   Status status;
   *cenv = nullptr;
-  // If underlying env is not defined, then use PosixEnv
-  if (!base_env) {
-    base_env = Env::Default();
+  // If underlying FileSystem is not defined, then use PosixFileSystem
+  auto fs = base_fs;
+  if (!fs) {
+    fs = FileSystem::Default();
   }
-  std::unique_ptr<AwsEnv> aenv(new AwsEnv(base_env, cloud_options, info_log));
+  std::unique_ptr<AwsEnv> aenv(new AwsEnv(fs, cloud_options, info_log));
+
+  auto env = aenv->NewCompositeEnvFromThis(Env::Default());
   ConfigOptions config_options;
-  config_options.env = aenv.get();
+  config_options.env = env.get();
   status = aenv->PrepareOptions(config_options);
   if (status.ok()) {
     *cenv = aenv.release();
   }
   return status;
 }
-#endif  // USE_AWS
 
-Status AwsEnv::NewAwsEnv(Env* env, std::unique_ptr<CloudEnv>* cenv) {
-#ifdef USE_AWS
-  cenv->reset(new AwsEnv(env, CloudEnvOptions()));
+Status AwsEnv::NewAwsEnv(const std::shared_ptr<FileSystem>& fs,
+                         std::unique_ptr<CloudEnv>* cenv) {
+  cenv->reset(new AwsEnv(fs, CloudEnvOptions()));
   return Status::OK();
-#else
-  (void) env;
-  cenv->reset();
-  return Status::NotSupported("AWS not supported");
-#endif // USE_AWS
 }
 
+#endif  // USE_AWS
```

Comment:
Did you mean to change `USE_AWS` here? I don't think linking will succeed with `USE_AWS` turned off if you don't define `AwsEnv::NewAwsEnv`.

Reply:
Oh I see, it's `CloudEnv::NewAwsEnv` (not `AwsEnv::NewAwsEnv`) that's defined in the public header file and that one is properly declared even when `USE_AWS` is off. All good.